### PR TITLE
Handle numerical overflow

### LINKED ISSUE #32: Support overflow correctly
Currently:
* 32 bit ops wrap around, but crab does not support that, so it's an havoc
* 64 bit ops do not wrap around. Havocing them is much too imprecise. Checking for UINT64_MAX is impossible since crab looks for an int, not a long long

### PR DESCRIPTION
**Pull request by elazarg**
Fix #32

### CODE REVIEW

File: src/crab_constraints.cpp [hunk 1]
```diff
@@ -327,16 +332,18 @@ machine_t::machine_t(variable_factory_t& vfac, program_info info)
  */
 void machine_t::setup_entry(basic_block_t& entry)
 {
-    auto& machine = *this;
+    machine_t& machine = *this;
     entry.havoc(machine.top);
-    entry.assign(machine.maxint64, INT64_MAX);
+    entry.assign(machine.maxint64, 1000000);
```
Comment (caballa) [inline src/crab_constraints.cpp]: Shouldn't be `MY_INT_MAX`

File: src/crab_constraints.cpp [hunk 2]
```diff
@@ -327,16 +332,18 @@ machine_t::machine_t(variable_factory_t& vfac, program_info info)
...
+    entry.assign(machine.minint64, -1000000);
```
Comment (caballa) [inline src/crab_constraints.cpp]: Same but `MY_INT_MIN`

File: src/crab_constraints.cpp (hunk 1 above)
Comment (elazarg) [reply src/crab_constraints.cpp]: Good catch

Comment (caballa) [reply src/crab_constraints.cpp]: I see, ok

**Comment by elazarg** [comment]
Running full benchmark now... It's gonna take a while.
- [ ] run blowup experiment 
- [ ] run on loops

**Comment by elazarg** [comment]
@caballa what numbers should we put, and how to do this without errors?
The maximum size of a region is 64k.

**Comment by caballa** [comment]
Ideally MAX should be 2^64 - 64k. Are we considering MIN=0 (i.e., can addressed be negative?)

**Comment by caballa** [comment]
we can also make it an user parameter so that we can play easily with that

**Comment by caballa** [comment]
The code looks goods to me. Basically given two blocks and edge between them B1 --> B2, if there is an operation in B1 that might overflow you convert it into: B1 ---> B1_overflow, B1--->B1_underflow, B1_overflow--->B2, B1_underflow--->B2, and B1--->B2. correct?

**Comment by caballa** [comment]
DIV and MOD can also overflow if denominator is -1 and numerator is MIN_INT

**Comment by elazarg** [comment]
> The code looks goods to me. Basically given two blocks and edge between them B1 --> B2, if there is an operation in B1 that might overflow you convert it into: B1 ---> B1_overflow, B1--->B1_underflow, B1_overflow--->B2, B1_underflow--->B2, and B1--->B2. correct?

Yes.

**Comment by elazarg** [comment]
benchmark passed. Same runtime for crabZone. Similar for others. More errors for polyhedra for some reason.

**Comment by caballa** [comment]
For poly you mean the vector exception error ?

**Comment by elazarg** [comment]
I did not see the errors. Maybe I will have a look at it later.

**Comment by elazarg** [comment]
Okay. Let's choose the right numbers and merge to master.
What do you propose?

**Comment by caballa** [comment]
Can you try MAX = 2^64 - 64k ?

**Comment by elazarg** [comment]
What shall I say. It does seem to work.
One last glance and I'll merge.

**Comment by caballa** [comment]
Too good to be true. Do we have some small tests that overflow? Just to make sure we can detect them.

**Comment by elazarg** [comment]
Good idea. I'm working on it. Though with optimizations enabled this is not trivial

**Comment by elazarg** [comment]
But why shouldn't it work? INT_MAX is far smaller than what we actually need, and close to what we already tried successfully.

**Comment by caballa** [comment]
No reason. Everything makes sense. Just wanted to be even more sure. But if you think it's too hard I think it's safe to merge

**Comment by elazarg** [comment]
Sanity check is important, but I don't have anything; my attempts showed havoc in overflow, but I don't have a consistent test, which is pity.

**Comment by elazarg** [comment]
Is this convincing? Just random arithmetic there. `r3` overflows and havoced:
```
{meta_size -> [-oo, 0], data_size -> [0, 1073741824], r0 -> [1, 225], t0 -> [-5, -5], r1 -> [4098, 2147418112], off1 -> [0, 0], t1 -> [-1, -1], r3 -> [1, 225], t2 -> [-5, -5], r2 -> [0, 1073741824], t3 -> [-5, -5], r2-data_size<=0, off1-data_size<=0, r3-r0<=0, r2-off1<=1073741824, data_size-off1<=1073741824, r0-r3<=0, data_size-r2<=0, off1-r2<=0}
5:2:
  assert(-t2 < 6) // line=5 column=0;
  assert(-t3 < 6) // line=5 column=0;
  r3 = r3*r2;
  t3 = -5;
  havoc(off3);
  goto 5:2:overflow,5:2:underflow,5:2:exit;

{meta_size -> [-oo, 0], data_size -> [0, 1073741824], r0 -> [1, 225], t0 -> [-5, -5], r1 -> [4098, 2147418112], off1 -> [0, 0], t1 -> [-1, -1], r3 -> [0, 241591910400], t3 -> [-5, -5], off1-data_size<=0, data_size-off1<=1073741824}

{meta_size -> [-oo, 0], data_size -> [0, 1073741824], r0 -> [1, 225], t0 -> [-5, -5], r1 -> [4098, 2147418112], off1 -> [0, 0], t1 -> [-1, -1], t3 -> [-5, -5], off1-data_size<=0, data_size-off1<=1073741824}
5:2:exit:
  goto 5:3;

{meta_size -> [-oo, 0], data_size -> [0, 1073741824], r0 -> [1, 225], t0 -> [-5, -5], r1 -> [4098, 2147418112], off1 -> [0, 0], t1 -> [-1, -1], t3 -> [-5, -5], off1-data_size<=0, data_size-off1<=1073741824}

{meta_size -> [-oo, 0], data_size -> [0, 1073741824], r0 -> [1, 225], t0 -> [-5, -5], r1 -> [4098, 2147418112], off1 -> [0, 0], t1 -> [-1, -1], r3 -> [0, 241591910400], t3 -> [-5, -5], off1-data_size<=0, data_size-off1<=1073741824}
5:2:overflow:
  assume(-r3 < -2147483647);
  havoc(r3);
  goto 5:2:exit;

{meta_size -> [-oo, 0], data_size -> [0, 1073741824], r0 -> [1, 225], t0 -> [-5, -5], r1 -> [4098, 2147418112], off1 -> [0, 0], t1 -> [-1, -1], t3 -> [-5, -5], off1-data_size<=0, data_size-off1<=1073741824}
```

**Comment by caballa** [comment]
Yes, it seems to work

**Comment by elazarg** [comment]
Thanks!

**Pull request merged**